examples/tut-6-rpc/client-py/index.py

```diff
@@ -0,0 +1,130 @@
+
+
+def mk_rpc(proto='http',
+           host=os.environ['REACH_RPC_SERVER'],
+           port=os.environ['REACH_RPC_PORT']):
+
+    def rpc(m, *args):
+        wait_for_port(port, host)
+        lab = 'RPC %s %s' % (m, json.dumps([*args]))
+        print(lab)
+        ans = requests.post('%s://%s:%s%s' % (proto, host, port, m),
+                            json=[*args])
+        print('%s ==> %s' % (lab, json.dumps(ans.json())))
+        return ans.json()
+
+    def rpc_callbacks(m, arg, cbacks):
+        vals  = {k: v    for k, v in cbacks.items() if not callable(v)}
+        meths = {k: True for k, v in cbacks.items() if     callable(v)}
+        p     = rpc(m, arg, vals, meths)
+
+        while True:
+            if p['t'] == 'Done':
+                return p
+
+            elif p['t'] == 'Kont':
+                cback = cbacks[p['m']]
+                ans   = cback(*p['args'])
+                p     = rpc('/kont', p['kid'], ans)
+
+            else:
+                # TODO fix swallowed exceptions
+                raise Exception('Illegal callback return: %s' % json.dumps(p))
+
+    return rpc, rpc_callbacks
+
+
+def main():
+    print('I am the client')
+
+    rpc, rpc_callbacks = mk_rpc()
+
+    starting_balance = rpc('/stdlib/parseCurrency', 10)
+    acc_alice        = rpc('/stdlib/newTestAccount', starting_balance)
+    acc_bob          = rpc('/stdlib/newTestAccount', starting_balance)
+
+    fmt          = lambda x: rpc('/stdlib/formatCurrency', x, 4)
+    get_balance  = lambda w: fmt(rpc('/stdlib/balanceOf', w))
+
+    before_alice = get_balance(acc_alice)
+    before_bob   = get_balance(acc_bob)
+
+    ctc_alice    = rpc('/acc/deploy', acc_alice)
+    ctc_bob      = rpc('/acc/attach', acc_bob, rpc('/ctc/getInfo', ctc_alice))
+
+    HAND         = ['Rock', 'Paper', 'Scissors']
+    OUTCOME      = ['Bob wins', 'Draw', 'Alice wins']
+
+    def getHand(who):
+        hand = random.randint(0, 2)  # TODO better source of randomness(?)
+        print('%s played %s' % (who, HAND[hand]))
+        return hand
+
+    acceptWager = lambda amt: print('Bob accepts the wager of %s' % fmt(amt))
+
+    player = lambda who: \
+        {'stdlib.hasRandom': True,
+         'getHand':       lambda:   getHand(who),
+         'informTimeout': lambda:   print('%s observed a timeout' % who),
+         'seeOutcome':    lambda n: print('%s saw outcome %s'
+            % (who, OUTCOME[rpc('/stdlib/bigNumberToNumber', n)]))
+         }
+
+    play_alice = lambda: rpc_callbacks(
+        '/backend/Alice',
+        ctc_alice,
+        dict(wager=rpc('/stdlib/parseCurrency', 5), **player('Alice')))
+
+    play_bob = lambda: rpc_callbacks(
+        '/backend/Bob',
+        ctc_bob,
+        dict(acceptWager=acceptWager, **player('Bob')))
+
+    with ThreadPoolExecutor() as ex:
+        t_alice = ex.submit(play_alice)
+        t_bob   = ex.submit(play_bob)
+
+        wait([t_alice, t_bob], return_when=ALL_COMPLETED)
```

Comment:
I am not an average Python programmer, but when I read this, I imagine that an average Python programmer thinks they need to understand more than

```
alice = threading.Thread(target=play_alice)
bob = threading.Thread(target=play_bob)
alice.start()
bob.start()
alice.join()
bob.join()
```

It would be nice if you could put the `start` in the constructor... but maybe I am trying to over-optimize when I really don't know what Pythonistas expect

Reply:
The `with` pattern has special significance and you see it used + promoted all over the place in Python but concurrency/parallelism have always been very weak points, so most people avoid the standard library's answer to these (using 3rd party packages instead) to build anything beyond small stuff (which is why it's hard to choose one standard library API as opposed to another).

I swapped out the threading code in 9ab5bef to match yours, and maybe that's easier for beginners to read anyway, but unfortunately you can't chain `start` or put it in the constructor; their API forces you to call methods of the `Thread` instances separately.